services/comparison/types.py

```diff
@@ -39,6 +39,11 @@ class Comparison(object):
     enriched_pull: EnrichedPull
     current_yaml: Optional[UserYaml] = None
 
+    # FIXME: The functions down below do not make sense given that we assume
+    # a `FullCommit` and its `report` to always exist.
```

Comment:
Hmmm, will changing this to a ReadOnlyReport cause issues then? What if we did `Report | ReadOnlyReport` and do checks in the fns below to do `if report and ...`? Although not sure if it goes beyond the scope of this PR

Reply:
Changing this to `ReadOnlyReport` is fixing wrong type annotations, as both the caller and downstream code already work with `ReadOnlyReport`.

The problem here is actually:
- Every caller always sets the `FullCommit`, so `self.head` won’t ever be `None`.
- On the other hand though, `FullCommit.report` could be `None` contrary to the declared type.
- Probably a bunch of downstream code was not properly checking `has_head_report`, and just assuming that `head.report` is never `None`, which is the part thats blowing up.

Reply:
Ahhhhh I see I see. Hmmmm yeah if there's a way we can spot/document where downstream code assumes head.report to be not None, that would help, but beyond the scope of this ticket. Is there a future issue to address this?

Reply:
I created https://github.com/codecov/engineering-team/issues/2917 to revisit this in the future.